### PR TITLE
Retain literal type for prefix plus on number literal

### LINKED ISSUE #17392: Prefix plus on number literal loses literal type
**TypeScript Version:**  2.4.0

**Code**

```ts
let x: 1 = +1;
```

**Expected behavior:**

It would compile.

**Actual behavior:**

Error:
>Type 'number' is not assignable to type '1'.

I can see why this is happening: in general, `+x` won't have the same type as `x`, since it's a cast to number. But `+1` is definitely the same as `1`! This is mostly just mildly surprising.

There are two solutions as I see:

* (a) for any number literal, `+x` has type `x`, rather than `number`
* or (b) for any expression `e` whose type `T` is a subset of `number`, `+e` has type `T`.

The latter is more general, but likely more dangerous. So I'm happy with either. Obviously, this isn't a showstopper.

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #17392

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16952,9 +16952,13 @@ namespace ts {
             if (operandType === silentNeverType) {
                 return silentNeverType;
             }
-            if (node.operator === SyntaxKind.MinusToken && node.operand.kind === SyntaxKind.NumericLiteral) {
+            const isOperandNumericLiteral = node.operand.kind === SyntaxKind.NumericLiteral;
+            if (isOperandNumericLiteral && node.operator === SyntaxKind.MinusToken) {
                 return getFreshTypeOfLiteralType(getLiteralType(-(<LiteralExpression>node.operand).text));
             }
+            if (isOperandNumericLiteral && node.operator === SyntaxKind.PlusToken) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you fold these into one check for `node.operand.kind === SyntaxKind.NumericLiteral`?

Comment (ikatyang) [inline src/compiler/checker.ts]: Sorry, I forgot I could write that way. 😅

**Comment by ikatyang** [comment]
@DanielRosenwasser done

**Comment by DanielRosenwasser** [comment]
Thanks @ikatyang

**Comment by DanielRosenwasser** [comment]
I guess this also fixed #15346.

**Pull request merged**